### PR TITLE
omdb could try harder to find instances that are online

### PR DESCRIPTION
**Pull request by davepacheco**
This seeks to address the problem mentioned [here](https://github.com/oxidecomputer/omicron/issues/5480#issuecomment-2075746007).  A real solution would involve more sophisticated connection management for all of our clients (a la cueball).  That's not a small project.  Here, I've thrown together kind of a hack that I think meaningfully improves the omdb experience.

### CODE REVIEW

**Comment by davepacheco** [comment]
I tested this manually on madrid.  For reference, there are currently three Nexus zones on four sleds there.

grep'ing `nexus` from the `omdb nexus blueprints show` output: 

```
    nexus             2d7cf6b7-c0f0-442c-a24c-4891e8b69a1f   in service   fd00:1122:3344:102::4  
    nexus             5cf52965-3fa2-4d86-8475-8061bb973705   in service   fd00:1122:3344:104::5 
    nexus             75710c69-f279-48b9-8f7c-8f9b1c89d251   in service   fd00:1122:3344:103::4  
```

From `omdb db sleds`:

```
# omdb db sleds
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:104::4]:32221,[fd00:1122:3344:102::3]:32221,[fd00:1122:3344:103::3]:32221,[fd00:1122:3344:104::3]:32221,[fd00:1122:3344:101::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (53.0.0)
SERIAL      IP                            ROLE     ID  
BRM44220001 [fd00:1122:3344:102::1]:12345 scrimlet 52d8505c-7b3e-4e0a-9df0-30dc6d146b06
BRM42220004 [fd00:1122:3344:101::1]:12345 -        5fb8f008-55e6-4719-b4b1-3a17d8c760ec
BRM42220081 [fd00:1122:3344:104::1]:12345 scrimlet a8e85656-7f84-4ce0-b66b-8ddcecbed46a
BRM42220046 [fd00:1122:3344:103::1]:12345 -        ca0d174e-e159-4447-8df6-126dd3137549
```

From `omdb db dns names internal 1`, we have these Nexus DNS names:

```
  _nexus._tcp                                        (records: 3)
      SRV  port 12221 2d7cf6b7-c0f0-442c-a24c-4891e8b69a1f.host.control-plane.oxide.internal
      SRV  port 12221 5cf52965-3fa2-4d86-8475-8061bb973705.host.control-plane.oxide.internal
      SRV  port 12221 75710c69-f279-48b9-8f7c-8f9b1c89d251.host.control-plane.oxide.internal
```

All we really need to know is that we have these three Nexus IPs in this order:

```
fd00:1122:3344:102::4
fd00:1122:3344:104::5
fd00:1122:3344:103::4
```

and which sleds those are on, which we can determine from the IP addresses (e.g., fd00:1122:3344:102::4 is on BRM44220001).

Except where I mention it below, every one of the `omdb` invocations below completed immediately.

First, with all three zones online:

```
root@oxz_switch1:~# /var/tmp/omdb-dap-connretry nexus blueprints list
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:102::4]:12221
T ENA ID                                   PARENT TIME_CREATED
* no  770c2a46-0b8c-4747-81b5-58882e6d4904 <none> 2024-04-24T00:11:13.251Z
```

We can see it picked the first one in DNS, as expected.  I disabled the first Nexus SMF instance using `svcadm -z  oxz_nexus_2d7cf6b7-c0f0-442c-a24c-4891e8b69a1f disable -s nexus` on `BRM44220001`.  This should have the effect of producing an ECONNREFUSED error.  omdb may report this (or may not, if one of the other connections gets established first) but it should very quickly succeed with one of the other IPs.  Indeed:

```
root@oxz_switch1:~# /var/tmp/omdb-dap-connretry nexus blueprints list
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
warning: failed to connect to Nexus at [fd00:1122:3344:102::4]:12221: connect "[fd00:1122:3344:102::4]:12221": Connection refused (os error 146)
note: using Nexus URL http://[fd00:1122:3344:104::5]:12221
T ENA ID                                   PARENT TIME_CREATED
* no  770c2a46-0b8c-4747-81b5-58882e6d4904 <none> 2024-04-24T00:11:13.251Z
```

There it used the second IP.  When I re-enable the SMF service, I get the same as the previous output (no warning).

Then I halted the first Nexus zone to simulate a more complete failure.  Here, we'd expect connect attempts to time out, not fail explicitly, since there's no netstack running for that IP.  I tested this with `zoneadm -z  oxz_nexus_2d7cf6b7-c0f0-442c-a24c-4891e8b69a1f halt` on `BRM44220001`.  Here, we'd expect omdb to work using one of the other IPs and _not_ produce a warning:

```
root@oxz_switch1:~# /var/tmp/omdb-dap-connretry nexus blueprints list
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:104::5]:12221
T ENA ID                                   PARENT TIME_CREATED    
* no  770c2a46-0b8c-4747-81b5-58882e6d4904 <none> 2024-04-24T00:11:13.251Z 
```

Success.  I booted the first zone again and omdb picked it up again:

```
root@oxz_switch1:~# /var/tmp/omdb-dap-connretry nexus blueprints list
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:102::4]:12221
T ENA ID                                   PARENT TIME_CREATED
* no  770c2a46-0b8c-4747-81b5-58882e6d4904 <none> 2024-04-24T00:11:13.251Z
```

Next I disabled the SMF service in the first two zones.  As expected, we get two warnings and connect to the third one:

```
root@oxz_switch1:~# /var/tmp/omdb-dap-connretry nexus blueprints list
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
warning: failed to connect to Nexus at [fd00:1122:3344:102::4]:12221: connect "[fd00:1122:3344:102::4]:12221": Connection refused (os error 146)
warning: failed to connect to Nexus at [fd00:1122:3344:104::5]:12221: connect "[fd00:1122:3344:104::5]:12221": Connection refused (os error 146)
note: using Nexus URL http://[fd00:1122:3344:103::4]:12221
T ENA ID                                   PARENT TIME_CREATED
* no  770c2a46-0b8c-4747-81b5-58882e6d4904 <none> 2024-04-24T00:11:13.251Z
```

(Note that we could as well not have seen the warnings, if we successfully connected to the third IP before the failures.)

Then I disabled the Nexus SMF service in all three zones:

```
root@oxz_switch1:~# /var/tmp/omdb-dap-connretry nexus blueprints list
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
warning: failed to connect to Nexus at [fd00:1122:3344:102::4]:12221: connect "[fd00:1122:3344:102::4]:12221": Connection refused (os error 146)
warning: failed to connect to Nexus at [fd00:1122:3344:104::5]:12221: connect "[fd00:1122:3344:104::5]:12221": Connection refused (os error 146)
warning: failed to connect to Nexus at [fd00:1122:3344:103::4]:12221: connect "[fd00:1122:3344:103::4]:12221": Connection refused (os error 146)
Error: failed to connect to any instances of Nexus
```

That's as expected, too.  Now let's turn it back on in all three:

```
root@oxz_switch1:~# /var/tmp/omdb-dap-connretry nexus blueprints list
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:102::4]:12221
T ENA ID                                   PARENT TIME_CREATED
* no  770c2a46-0b8c-4747-81b5-58882e6d4904 <none> 2024-04-24T00:11:13.251Z
```

Now let's halt the first two zones:

```
root@oxz_switch1:~# /var/tmp/omdb-dap-connretry nexus blueprints list
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:103::4]:12221
T ENA ID                                   PARENT TIME_CREATED
* no  770c2a46-0b8c-4747-81b5-58882e6d4904 <none> 2024-04-24T00:11:13.251Z
```

Good.  Finally, let's halt the third.  This is the only invocation that _didn't_ complete right away, and that's expected, too.  omdb had to wait for all of these connections to time out, which took a couple of minutes:

```
# /var/tmp/omdb-dap-connretry nexus blueprints list
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
warning: failed to connect to Nexus at [fd00:1122:3344:102::4]:12221: connect "[fd00:1122:3344:102::4]:12221": No route to host (os error 148)
warning: failed to connect to Nexus at [fd00:1122:3344:103::4]:12221: connect "[fd00:1122:3344:103::4]:12221": No route to host (os error 148)
warning: failed to connect to Nexus at [fd00:1122:3344:104::5]:12221: connect "[fd00:1122:3344:104::5]:12221": No route to host (os error 148)
Error: failed to connect to any instances of Nexus
```

File: dev-tools/omdb/src/bin/omdb/main.rs [hunk 1]
```diff
@@ -129,6 +134,65 @@ impl Omdb {
...
+            "expected at least one address from successful DNS lookup for {:?}",
+            service_name
+        );
+
+        // The caller is going to pick one of these addresses to connect to.
+        // Let's try to pick one that's at least not obviously broken by
+        // attempting to connect to whatever we found and returning any that we
+        // successfully connected to.  It'd be nice if we could return the
+        // socket directly, but our callers are creating reqwest clients that
+        // cannot easily consume a socket directly.
```
Comment (sunshowers) [inline dev-tools/omdb/src/bin/omdb/main.rs:156]: I had this question but you answered it right here :)

**Review by sunshowers — APPROVED** [review]
I think this is pretty reasonable!

**Comment by davepacheco** [comment]
Thanks for the review!

**Review by jgallagher — APPROVED** [review]

**Pull request merged**